pkg/controllers/provisioning/v1alpha1/reallocation/annotate.go

```diff
@@ -0,0 +1,97 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reallocation
+
+import (
+	"context"
+	"fmt"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	utilsnode "github.com/awslabs/karpenter/pkg/utils/node"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type Annotator struct {
+	kubeClient client.Client
+}
+
+// Takes in a list of underutilized nodes and populates the nodes with the ProvisionerUnderutilized Label
+func (a *Annotator) LabelUnderutilized(ctx context.Context, nodes []*v1.Node) {
+	for _, node := range nodes {
+		persisted := node.DeepCopy()
+		node.Labels[v1alpha1.ProvisionerUnderutilizedKey] = "true"
+		if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			zap.S().Debugf("Continuing after failing to patch node %s, %s", node.Name, err.Error())
+		}
+	}
+}
+
+// AddTTLs takes in a list of underutilized nodes and adds TTL to them
+func (a *Annotator) AddTTLs(ctx context.Context, nodes []*v1.Node) error {
+	for _, node := range nodes {
+		persisted := node.DeepCopy()
+		node.Annotations[v1alpha1.ProvisionerTTLKey] = time.Now().Add(300 * time.Second).Format(time.RFC3339)
+		if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			return fmt.Errorf("patching node, %w", err)
+		}
+		zap.S().Debugf("Added TTL to node %s", node.Name)
+	}
+	return nil
+}
+
+// RemoveUtilizedTTLs takes in a list of nodes labeled underutilized and removes TTL if there is sufficient resource usage
+func (a *Annotator) RemoveUtilizedTTLs(ctx context.Context, nodes []*v1.Node) {
+	for _, node := range nodes {
+		pods, err := a.getPodsOnNode(ctx, node.Name)
+		if err != nil {
+			zap.S().Debugf("Ignoring error: could not get pods on node %s, %s", node.Name, err.Error())
+			continue
+		}
+		if !utilsnode.IsUnderutilized(pods) {
+			persisted := node.DeepCopy()
+			delete(node.Labels, v1alpha1.ProvisionerUnderutilizedKey)
+			delete(node.Annotations, v1alpha1.ProvisionerTTLKey)
+			if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+				zap.S().Debugf("Continuing after failing to remove TTL from node %s, %s", node.Name, err.Error())
+			} else {
+				zap.S().Debugf("Removed TTL from node %s", node.Name)
+			}
+		}
+	}
+}
+
+// CordonNodes takes in a list of expired nodes as input and cordons them
+func (a *Annotator) CordonNodes(ctx context.Context, nodes []*v1.Node) error {
+	for _, node := range nodes {
+		persisted := node.DeepCopy()
+		node.Spec.Unschedulable = true
+		if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			return fmt.Errorf("patching node %s, %w", node.Name, err)
+		}
+		zap.S().Debugf("Cordoned node %s", node.Name)
+	}
+	return nil
+}
+
+// Get Pods scheduled to a node
+func (a *Annotator) getPodsOnNode(ctx context.Context, nodeName string) (*v1.PodList, error) {
```

Comment:
I might've already said this, but why doesn't this just return a slice of pods, rather than exposing the PodList concept?

Reply:
This is an exact copy of the function in `filter.go` but I replicated it here, since I needed it. I can change it to result in a `[]*v1.Pod`, as that's the pattern I'm doing in other functions like in `filter.go`

Reply:
Factor your code to avoid copying code. Whenever you copying code, it should be a red flag that something isn't designed well.